File: Python/Problem_056.py
def multiply_digits(digits: list[int], m: int) -> None:
    carry = 0
    for i in range(len(digits)):
        prod = digits[i] * m + carry
        digits[i] = prod % 10
        carry = prod // 10
    while carry:
        digits.append(carry % 10)
        carry //= 10

def digit_sum(digits: list[int]) -> int:
    return sum(digits)

def euler_56() -> int:
    best = 0
    for a in range(1, 100):
        for b in range(1, 100):
            digits = [1]  # represents 1
            for _ in range(b):
                multiply_digits(digits, a)
            best = max(best, digit_sum(digits))
    return best

if __name__ == "__main__":
    print(euler_56())  # 972

File: Java/Problem_056.java
import java.util.ArrayList;

public class Euler56 {

    static void multiplyDigits(ArrayList<Integer> digits, int m) {
        int carry = 0;
        for (int i = 0; i < digits.size(); i++) {
            int prod = digits.get(i) * m + carry;
            digits.set(i, prod % 10);
            carry = prod / 10;
        }
        while (carry > 0) {
            digits.add(carry % 10);
            carry /= 10;
        }
    }

    static int digitSum(ArrayList<Integer> digits) {
        int s = 0;
        for (int d : digits) s += d;
        return s;
    }

    static int solve() {
        int best = 0;
        for (int a = 1; a < 100; a++) {
            for (int b = 1; b < 100; b++) {
                ArrayList<Integer> digits = new ArrayList<>();
                digits.add(1);
                for (int i = 0; i < b; i++) multiplyDigits(digits, a);
                best = Math.max(best, digitSum(digits));
            }
        }
        return best;
    }

    public static void main(String[] args) {
        System.out.println(solve()); // 972
    }
}
